File: python-sdk/src/projectledger/receipts/canonicalize.py
"""RFC 8785 JSON Canonicalization Scheme (JCS).

Pure-Python implementation. Matches the output of any compliant JCS
implementation byte-for-byte, including the TypeScript reference SDK's
output.

Rules implemented:
  - JSON object members are sorted by key in UTF-16 code-unit order.
  - No insignificant whitespace.
  - Numbers serialized per ECMAScript Number::toString base 10
    (ECMA-262 7.1.12.1), matching RFC 8785 §3.2.2.3.
  - Strings escaped per RFC 8785 §3.2.2.2 (JSON.stringify semantics):
    only ", \\, and the C0 controls are escaped; <>&/ and all
    non-ASCII (including DEL 0x7F, U+2028, U+2029) are emitted raw.
"""

from __future__ import annotations

from typing import Any

# Characters with dedicated short escapes (RFC 8785 §3.2.2.2).
_SHORT_ESCAPES = {
    0x08: "\\b",
    0x09: "\\t",
    0x0A: "\\n",
    0x0C: "\\f",
    0x0D: "\\r",
    0x22: '\\"',
    0x5C: "\\\\",
}


def _escape_string(s: str) -> str:
    """Serialize a JSON string value per ECMAScript JSON.stringify.

    Escapes ONLY: " \\ and the C0 controls (< 0x20). Everything else --
    including < > & /, DEL 0x7F, U+2028, U+2029, and all non-ASCII -- is
    emitted raw (encoded as UTF-8 at the byte layer). Short escapes are
    used for \\b \\f \\n \\r \\t; other C0 controls use \\u00XX (lowercase
    hex, 4 digits).
    """
    out = ['"']
    for ch in s:
        code = ord(ch)
        esc = _SHORT_ESCAPES.get(code)
        if esc is not None:
            out.append(esc)
        elif code < 0x20:
            out.append("\\u%04x" % code)
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _format_number(x: float) -> str:
    """Format a finite double per ECMAScript Number::toString base 10.

    Implements ECMA-262 7.1.12.1 steps 6-10. Uses Python ``repr`` to obtain
    the shortest round-tripping decimal (mantissa digits ``s`` and exponent
    giving ``n`` where value == s x 10^(n-k), k == len(s)), then applies the
    range rules.
    """
    if x != x or x in (float("inf"), float("-inf")):
        raise ValueError("Non-finite numbers not allowed in JCS")

    if x == 0:
        # Covers +0 and -0.
        return "0"

    if x < 0:
        return "-" + _format_number(-x)

    # Obtain shortest round-trip decimal from repr, e.g. '0.1', '1234.5',
    # '1e-07', '1e+21', '5e-324', '100000000000000000000'.
    r = repr(x)

    # Split mantissa and exponent.
    if "e" in r or "E" in r:
        mant, _, exp_str = r.replace("E", "e").partition("e")
        exp = int(exp_str)
    else:
        mant = r
        exp = 0

    # Split mantissa into integer and fractional parts.
    if "." in mant:
        int_part, frac_part = mant.split(".", 1)
    else:
        int_part, frac_part = mant, ""

    # Build the significant-digit string s and the decimal exponent n such
    # that value == s x 10^(n - k), where k = len(s).
    digits = int_part + frac_part
    # Position of the decimal point measured from the left of `digits`:
    # int_part contributes len(int_part) integer digits, then `exp` shifts it.
    point_pos = len(int_part) + exp

    # Strip leading zeros (adjusting the point position accordingly).
    lead = 0
    while lead < len(digits) - 1 and digits[lead] == "0":
        lead += 1
    digits = digits[lead:]
    point_pos -= lead

    # Strip trailing zeros (these do not affect point_pos).
    digits = digits.rstrip("0")
    if digits == "":
        # Value collapsed to zero (shouldn't happen since x != 0).
        return "0"

    s = digits
    k = len(s)
    n = point_pos

    # Apply ECMA-262 7.1.12.1 range rules.
    if k <= n <= 21:
        # Step 6: s followed by (n - k) zeros.
        return s + "0" * (n - k)
    if 0 < n <= 21:
        # Step 7: first n digits, '.', remaining k - n digits.
        return s[:n] + "." + s[n:]
    if -6 < n <= 0:
        # Step 8: '0.', (-n) zeros, then s.
        return "0." + "0" * (-n) + s
    # Step 9/10: exponential form.
    e = n - 1
    if e >= 0:
        exp_out = "e+" + str(e)
    else:
        exp_out = "e-" + str(-e)
    if k == 1:
        return s + exp_out
    return s[0] + "." + s[1:] + exp_out


def _serialize(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int):
        # Integers print with no decimal point and no exponent.
        return str(value)
    if isinstance(value, float):
        return _format_number(value)
    if isinstance(value, str):
        return _escape_string(value)
    if isinstance(value, list):
        parts = [_serialize(v) for v in value]
        return "[" + ",".join(parts) + "]"
    if isinstance(value, dict):
        # RFC 8785 §3.2.3: sort by UTF-16 code units.
        items = sorted(value.items(), key=lambda kv: kv[0].encode("utf-16-be"))
        parts = [f"{_escape_string(k)}:{_serialize(v)}" for k, v in items]
        return "{" + ",".join(parts) + "}"
    raise TypeError(f"Cannot canonicalize value of type {type(value).__name__}")


def canonicalize(value: Any) -> str:
    """Return the RFC 8785 canonical string form of `value`."""
    return _serialize(value)


def canonicalize_bytes(value: Any) -> bytes:
    """Return the RFC 8785 canonical bytes (UTF-8) of `value`."""
    return canonicalize(value).encode("utf-8")

File: java-sdk/src/main/java/io/projectledger/receipts/Canonicalize.java
package io.projectledger.receipts;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

/**
 * RFC 8785 JSON Canonicalization Scheme.
 *
 * Object keys sorted by UTF-16 code units. No insignificant whitespace.
 * Numbers serialized via Jackson (matches ECMAScript for the receipts
 * schema's integer-only numeric fields).
 *
 * Output is byte-identical to the TypeScript reference SDK.
 */
public final class Canonicalize {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private Canonicalize() {}

    /** Return the canonical string form of an arbitrary value. */
    public static String canonicalize(Object value) {
        StringBuilder sb = new StringBuilder();
        write(sb, MAPPER.valueToTree(value));
        return sb.toString();
    }

    /** Return canonical bytes (UTF-8). */
    public static byte[] canonicalizeBytes(Object value) {
        return canonicalize(value).getBytes(StandardCharsets.UTF_8);
    }

    private static void write(StringBuilder sb, JsonNode node) {
        if (node == null || node.isNull()) {
            sb.append("null");
        } else if (node.isBoolean()) {
            sb.append(node.booleanValue() ? "true" : "false");
        } else if (node.isNumber()) {
            // RFC 8785 §3.2.2.3: numbers use the ECMAScript Number::toString
            // (base 10) algorithm on the IEEE-754 double value.
            sb.append(numberToString(node.asDouble()));
        } else if (node.isTextual()) {
            writeString(sb, node.textValue());
        } else if (node.isArray()) {
            sb.append('[');
            for (int i = 0; i < node.size(); i++) {
                if (i > 0) sb.append(',');
                write(sb, node.get(i));
            }
            sb.append(']');
        } else if (node.isObject()) {
            List<String> keys = new ArrayList<>();
            Iterator<String> it = node.fieldNames();
            while (it.hasNext()) keys.add(it.next());
            keys.sort(Canonicalize::compareUtf16);
            sb.append('{');
            for (int i = 0; i < keys.size(); i++) {
                if (i > 0) sb.append(',');
                writeString(sb, keys.get(i));
                sb.append(':');
                write(sb, node.get(keys.get(i)));
            }
            sb.append('}');
        }
    }

    private static int compareUtf16(String a, String b) {
        int al = a.length();
        int bl = b.length();
        int n = Math.min(al, bl);
        for (int i = 0; i < n; i++) {
            char ca = a.charAt(i);
            char cb = b.charAt(i);
            if (ca != cb) return Character.compare(ca, cb);
        }
        return Integer.compare(al, bl);
    }

    /**
     * Escape a JSON string exactly as ECMAScript JSON.stringify / RFC 8785:
     * escape only ", \\, and C0 controls (< 0x20) with the short forms
     * \b \f \n \r \t or a 6-char backslash-u escape (lowercase hex).
     * Everything else — including
     * < > & /, DEL 0x7F, U+2028, U+2029 and all non-ASCII — is emitted raw.
     */
    private static void writeString(StringBuilder sb, String s) {
        sb.append('"');
        int len = s.length();
        for (int i = 0; i < len; i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\b': sb.append("\\b"); break;
                case '\f': sb.append("\\f"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:
                    if (c < 0x20) {
                        sb.append("\\u");
                        String hex = Integer.toHexString(c);
                        for (int p = hex.length(); p < 4; p++) sb.append('0');
                        sb.append(hex);
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append('"');
    }

    /**
     * ECMAScript Number::toString (base 10), RFC 8785 §3.2.2.3. Ported from
     * the cyberphone/json-canonicalization NumberToJSON algorithm: derive the
     * shortest round-tripping (digits, exponent) pair from Double.toString and
     * apply the ES2015 7.1.12.1 range rules (steps 6–10).
     */
    static String numberToString(double value) {
        if (Double.isNaN(value) || Double.isInfinite(value)) {
            throw new IllegalArgumentException("Non-finite number not representable in JCS: " + value);
        }
        // -0 and +0 both format as "0".
        if (value == 0.0) {
            return "0";
        }
        StringBuilder result = new StringBuilder();
        if (value < 0) {
            result.append('-');
            value = -value;
        }

        // Derive the shortest round-tripping decimal as a significant-digit
        // string `s` and integer `n` such that the value equals
        // s * 10^(n - k), where k = s.length(). We find the fewest significant
        // digits (1..17) whose scientific formatting parses back to the exact
        // same double. This is guaranteed shortest and round-tripping — more
        // robust than Double.toString, which is not always shortest (e.g. the
        // min subnormal prints as "4.9E-324" though "5E-324" round-trips).
        String sci = null;
        for (int prec = 0; prec <= 16; prec++) {
            String cand = String.format(java.util.Locale.ROOT, "%." + prec + "e", value);
            if (Double.parseDouble(cand) == value) {
                sci = cand;
                break;
            }
        }
        if (sci == null) {
            // Fallback (should not happen): 17 significant digits always suffice.
            sci = String.format(java.util.Locale.ROOT, "%.16e", value);
        }

        // sci looks like "d.dddde[+-]XX" (or "de[+-]XX" for prec 0). Split it.
        int eIdx = sci.indexOf('e');
        String mantissa = sci.substring(0, eIdx);
        int sciExp = Integer.parseInt(sci.substring(eIdx + 1));
        // mantissa = d0[.d1d2...]; combine into a single digit string.
        String digits = mantissa.replace(".", "");
        // Value == digits * 10^(sciExp - (digits.length() - 1)).
        // So with s == digits and k == digits.length(), n == sciExp + 1.
        // Strip trailing zeros from `digits` (keep >= 1 digit); n stays the same
        // because dropping low-order zeros only removes least-significant digits.
        int end = digits.length();
        while (end > 1 && digits.charAt(end - 1) == '0') {
            end--;
        }
        String s = digits.substring(0, end);

        int k = s.length();
        int n = sciExp + 1; // value == s * 10^(n - k)

        if (k <= n && n <= 21) {
            // Rule 6: digits followed by (n - k) zeros.
            result.append(s);
            for (int i = 0; i < n - k; i++) result.append('0');
        } else if (0 < n && n <= 21) {
            // Rule 7: first n digits, ".", remaining k - n digits.
            result.append(s, 0, n).append('.').append(s, n, k);
        } else if (-6 < n && n <= 0) {
            // Rule 8: "0.", (-n) zeros, then s.
            result.append("0.");
            for (int i = 0; i < -n; i++) result.append('0');
            result.append(s);
        } else {
            // Rules 9/10: exponential.
            int e = n - 1;
            if (k == 1) {
                result.append(s);
            } else {
                result.append(s.charAt(0)).append('.').append(s, 1, k);
            }
            result.append('e');
            if (e >= 0) {
                result.append('+').append(e);
            } else {
                result.append('-').append(-e);
            }
        }
        return result.toString();
    }
}
